input/chainsync/chainsync.go

```diff
@@ -403,3 +420,70 @@ func (c *ChainSync) updateStatus(
 		c.statusUpdateFunc(*(c.status))
 	}
 }
+
+func getKupoClient(c *ChainSync) (*kugo.Client, error) {
+	k := kugo.New(kugo.WithEndpoint(c.kupoUrl))
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	healthUrl := strings.TrimRight(c.kupoUrl, "/") + "/v1/health"
+	req, err := http.NewRequestWithContext(ctx, "GET", healthUrl, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create health check request: %w", err)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to perform health check: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("health check failed with status code: %d", resp.StatusCode)
+	}
+
+	return k, nil
+}
+
+// resolveTransactionInputs resolves the transaction inputs by using the
+// Kupo client and fetching the corresponding transaction outputs.
+func resolveTransactionInputs(transaction ledger.Transaction, c *ChainSync) ([]ledger.TransactionOutput, error) {
+	var resolvedInputs []ledger.TransactionOutput
+
+	// Use Kupo client to resolve inputs if available
+	if c.kupoUrl != "" {
+		k, err := getKupoClient(c)
```

Comment:
Get k from kugo directly and skip the entire getKupoClient so you're not sending a health check request with each event.